packages/expo-notifications/ios/EXNotifications/Notifications/Emitter/EmitterModule.swift

```diff
@@ -0,0 +1,58 @@
+//  Copyright © 2024 650 Industries. All rights reserved.
+
+import ExpoModulesCore
+import UIKit
+import MachO
+
+let onDidReceiveNotification = "onDidReceiveNotification"
+let onDidReceiveNotificationResponse = "onDidReceiveNotificationResponse"
+let onDidClearNotificationResponse = "onDidClearNotificationResponse"
+
+public class EmitterModule: Module, NotificationDelegate {
+  public func definition() -> ModuleDefinition {
+    Name("ExpoNotificationsEmitter")
+
+    Events([onDidReceiveNotification, onDidReceiveNotificationResponse, onDidClearNotificationResponse])
+
+    OnStartObserving {
+      NotificationCenterManager.shared.addDelegate(self)
+    }
+
+    OnStopObserving {
+      NotificationCenterManager.shared.removeDelegate(self)
+    }
+
+    AsyncFunction("getLastNotificationResponseAsync") {(promise: Promise) in
+      if let lastResponse: UNNotificationResponse = NotificationCenterManager.shared.lastResponse {
+        promise.resolve(EXNotificationSerializer.serializedNotificationResponse(lastResponse))
+      }
+      promise.resolve(nil)
+    }
+
+    AsyncFunction("clearLastNotificationResponseAsync") {(promise: Promise) in
+      NotificationCenterManager.shared.lastResponse = nil
+      promise.resolve(nil)
+    }
+  }
+
+  public func didReceive(_ notification: UNNotification, completionHandler: @escaping (UIBackgroundFetchResult) -> Void) -> Bool {
+    completionHandler(.noData)
+    return true
+  }
+
+  public func didReceive(_ response: UNNotificationResponse, completionHandler: @escaping () -> Void) -> Bool {
+    NotificationCenterManager.shared.lastResponse = response
+    // TODO: convert serialization to Records
+    let serializedResponse = EXNotificationSerializer.serializedNotificationResponse(response)
+    self.sendEvent(onDidReceiveNotificationResponse, serializedResponse as [String: Any])
+    return true
+  }
+
+  public func willPresent(_ notification: UNNotification, completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) -> Bool {
+    // TODO: convert serialization to Records
+    let serializedNotification = EXNotificationSerializer.serializedNotification(notification)
+    self.sendEvent(onDidReceiveNotification, serializedNotification as [String: Any])
+    completionHandler([])
```

Comment:
```suggestion
```

AFAICT, the presence of this line means that all calls to https://docs.expo.dev/versions/latest/sdk/notifications/#schedulenotificationasyncrequest are effectively ignored. This is because the `completionHandler([])` says "don't do anything, ignore the notification".

I'm not sure how the old implementation functioned, but this one doesn't seem to work.

Reply:
This code runs only if the app has a listener registered to handle incoming notifications. The idea is that the app's UI will do any presentation necessary, and we should not also present an alert. The old implementation did the same thing.

Reply:
The old implementation calls the completion handler in the same way. https://github.com/expo/expo/blob/adb33c6c55f43d7ebd570b9f589adc6822a77b20/packages/expo-notifications/ios/EXNotifications/Notifications/Emitter/EXNotificationsEmitter.m#L94